File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> list = new LinkedList<Integer>();
    Stack<TreeNode> stack = new Stack<TreeNode>();
    public List<Integer> inorderTraversal(TreeNode root) {
        if (root == null) {
            return list;
        }
        TreeNode current = root;
        while (current != null || !stack.isEmpty()) {
            
            while(current != null) {
                stack.push(current);
                current = current.left;
            }
            if (current == null && stack.isEmpty()) {
                break;
            } 
            current = stack.pop();
            list.add(current.val);
            current=current.right;
                
        }
        return list;
    }
}

File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.cpp
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> inorder;
        stack<TreeNode*> nodes;
        
        if(root == NULL) return inorder;
        
        while(root != NULL || !nodes.empty()){
            //push left children if available
            while(root != NULL){
                nodes.push(root);
                root = root->left;
            }
            
            //retrieve top node and store its right child if exists
            root = nodes.top();
            nodes.pop();
            
            inorder.push_back(root->val);
            root = root->right;
        }
        
        return inorder;
    }
};